### PR TITLE
[Android] Fix game banners

### PR DESCRIPTION
**Pull request by mahdihijazi**
Use Bitmap.setPixels() instead of Bitmap.copyPixelsFromBuffer() the former
use non pre-multiplied values of the colors which is what we expect to
come from the native code.

### CODE REVIEW

**Comment by JosJuice** [comment]
This does fix the color issue for me, but I don't understand why. The alpha channel is always 0xFF, so why would it matter if the colors are treated as pre-multiplied or not?

But either way, I think the change that this PR makes is good regardless of whether it fixes the color issue.

**Comment by mahdihijazi** [comment]
@JosJuice me too, I think I am missing something because according to the docs [setPixels](https://goo.gl/ad67NV) should work similar to [copyPixelsFromBuffer](https://developer.android.com/reference/android/graphics/Bitmap.html#copyPixelsFromBuffer(java.nio.Buffer)) except for the multiplication thing, and I did the multiplication manually and tried copyPixelsFromBuffer on the multiplied values and it didn't work.
Calling setPixels resulted in setting the correct colors for reason I am not sure about. I think some internal behavior of those methods is no clear in the docs. The only thing that is clear is that setPixels suppose to work with the non premultiplied values which should not matter since we have 0xFF alpha!

**Comment by MayImilae** [comment]
@mahdihijazi Is it possible that since the old code converted it to integers, it caused an endian swap which changed the colors?

**Comment by ds84182** [comment]
https://developer.android.com/reference/android/graphics/Bitmap.html#setPremultiplied(boolean)

> Bitmaps are always treated as pre-multiplied by the view system
and Canvas for performance reasons. Storing un-pre-multiplied data in a
Bitmap (through setPixel(int, int, int), setPixels(int[], int, int, int,
int, int, int), or BitmapFactory.Options.inPremultiplied) can lead to
incorrect blending if drawn by the framework.

https://developer.android.com/reference/android/graphics/Bitmap.html#setPixels(int[],
int, int, int, int, int, int)

> Replace pixels in the bitmap with the colors in the array. Each element
in the array is a packed int representing a non-premultiplied ARGB Color.

https://developer.android.com/reference/android/graphics/Bitmap.html#copyPixelsFromBuffer(java.nio.Buffer)

> Copy the pixels from the buffer, beginning at the current position,
overwriting the bitmap's pixels. The data in the buffer is not changed in
any way (unlike setPixels(), which converts from unpremultipled 32bit to
whatever the bitmap's native format is.

TL;DR: Bitmaps are always premultiplied, you shouldn't turn off
premultiplication, setPixels will always convert from unmultiplied ARGB to
the Bitmap's format, while the copyPixel*Buffer methods depend on the
Bitmap's pixel format and expects premultiplied data if the Bitmap is
premultiplied.

**Comment by mahdihijazi** [comment]
> TL;DR: Bitmaps are always premultiplied, you shouldn't turn off
premultiplication, setPixels will always convert from unmultiplied ARGB to
the Bitmap's format, while the copyPixel*Buffer methods depend on the
Bitmap's pixel format and expects premultiplied data if the Bitmap is
premultiplied.

How do you explain copyPixel*Buffer not working if we always use 0xFF alpha? after all it should not matter if we used non-premultiplied data if the alpha is always 0xFF

**Comment by ds84182** [comment]
Oh, thats because Skia's ARGB might not be actual ARGB, depending on graphics acceleration. [See here.](https://github.com/google/skia/blob/255095866b8c58f38e59e53cf0fa08c06a179636/include/core/SkPostConfig.h#L144) It was probably historically ARGB until Skia gained the OpenGL backend. copyPixelsFromBuffer does warn that it just does a plain copy, and that setPixels will automatically convert into the Bitmap's native format. The documentation for Bitmap.Config.ARGB_8888 doesn't specify an ordering: `Each pixel is stored on 4 bytes. Each channel (RGB and alpha for translucency) is stored with 8 bits of precision (256 possible values.) This configuration is very flexible and offers the best quality. It should be used whenever possible.`

**Comment by MayImilae** [comment]
Okaaaay... Can anyone explain to me why this worked? Ideally in english, not code!

**Comment by mahdihijazi** [comment]
@MayImilae Android view components always expect Bitmaps with premultiplied data inside, and since the Bitmap I am using is expecting premultiplied data, we have to use setPixels to convert the non  premultiplied data we receive from the native code into the expected format.  copyPixelsFromBuffer is not working because it just copy the data without changing the format to what the Bitmap expect.
